pkg/kv/kvclient/rangefeed/rangefeed_external_test.go

```diff
@@ -1645,3 +1646,210 @@ func (c *channelSink) Send(e *kvpb.RangeFeedEvent) error {
+		// We must enable desired scheduler settings before we start cluster,
+		// otherwise we will trigger processor restarts later and this test can't
+		// handle duplicated events.
+		kvserver.RangeFeedUseScheduler.Override(ctx, &settings.SV, s.useScheduler)
+		srv, _, db := serverutils.StartServer(t, base.TestServerArgs{
+			Settings: settings,
+		})
+		defer srv.Stopper().Stop(ctx)
+		ts := srv.ApplicationLayer()
+
+		scratchKey := append(ts.Codec().TenantPrefix(), keys.ScratchRangeMin...)
+		_, _, err := srv.SplitRange(scratchKey)
+		require.NoError(t, err)
+		scratchKey = scratchKey[:len(scratchKey):len(scratchKey)]
+		mkKey := func(k string) roachpb.Key {
+			return encoding.EncodeStringAscending(scratchKey, k)
+		}
+
+		sp := roachpb.Span{
+			Key:    scratchKey,
+			EndKey: scratchKey.PrefixEnd(),
+		}
+
+		// Enable rangefeeds, otherwise the thing will retry until they are enabled.
+		for _, l := range []serverutils.ApplicationLayerInterface{ts, srv.SystemLayer()} {
+			// Whether rangefeeds are enabled is a property of both the
+			// storage layer and the application layer.
+			kvserver.RangefeedEnabled.Override(ctx, &l.ClusterSettings().SV, true)
+		}
+
+		f, err := rangefeed.NewFactory(ts.AppStopper(), db, ts.ClusterSettings(), nil)
+		require.NoError(t, err)
+
+		metadata := make(chan *kvpb.RangeFeedMetadata)
+		initialScanDone := make(chan struct{})
+		rows := make(chan *kvpb.RangeFeedValue)
+		r, err := f.RangeFeed(ctx, "test", []roachpb.Span{sp}, ts.DB().Clock().Now(),
+			func(ctx context.Context, value *kvpb.RangeFeedValue) {
+				select {
+				case rows <- value:
+				case <-ctx.Done():
+				}
+			},
+			rangefeed.WithInitialScan(func(ctx context.Context) {
+				close(initialScanDone)
+			}),
+			rangefeed.WithOnMetadata(func(ctx context.Context, value *kvpb.RangeFeedMetadata) {
+				select {
+				case metadata <- value:
+				case <-ctx.Done():
+					return
+				}
+			}),
+		)
+		require.NoError(t, err)
+		defer r.Close()
+		{
+			// First meta msg for the new rangefeed.
+			meta := <-metadata
+			t.Logf("initial span %s-%s", meta.Span.Key, meta.Span.EndKey)
+			require.Equal(t, false, meta.FromManualSplit)
+			require.Equal(t, sp.Key, meta.Span.Key)
+			require.Equal(t, sp.EndKey, meta.Span.EndKey)
+		}
+		<-initialScanDone
+
+		// Confirm the rangefeed is running on the expected range.
+		require.NoError(t, db.Put(ctx, mkKey("c"), 2))
+		val := <-rows
+		require.Equal(t, mkKey("c"), val.Key)
+
+		splitKey := mkKey("b")
+		_, _, err = srv.SplitRange(splitKey)
+		require.NoError(t, err)
+		for {
+			// Expect a metadata event from the manual split.
+			meta := <-metadata
+			t.Logf("expected manual split new range key span %s-%s; manual split %t", meta.Span.Key, meta.Span.EndKey, meta.FromManualSplit)
+			require.Equal(t, true, meta.FromManualSplit)
+			if !meta.Span.EndKey.Equal(sp.EndKey) {
+				// New Rangefeed for LHS.
+				require.Equal(t, splitKey, meta.Span.EndKey)
+				require.Equal(t, sp.Key, meta.Span.Key)
```

Comment:
This test points out something interesting. Namely that the client is going to see two metadata events with FromManualSplit set to true (the LHS and RHS).  Out of curiosity, what's our plan on the PCR-side for this. Issue a split at both?

Reply:
Yeah, I think a nice to have would be to send the parent rangefeed start key over the wire. Then PCR could ignore sending splits if the new range's start key equals the parent rangefeed's start key.  This could avoid sending noop admin splits. 

We could also add a bool to the proto along the lines of `firstRangeFromRDIIteration`, but that seems hard to explain.

Before adding this additional plumbing I wanted to make sure everything worked.

Reply:
@stevendanna I added an extra commit that plumbs the previous start key. lemme know what you think!

Reply:
Looks reasonable to me!